File: internal/hook/hook.go
package hook

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/JustinDFuller/ban-code-comments/internal/languages"
	"github.com/JustinDFuller/ban-code-comments/internal/model"
	"github.com/JustinDFuller/ban-code-comments/internal/scanner"
)

type Mode string

const (
	ModeHard Mode = "hard"
	ModeWarn Mode = "warn"
)

type Event struct {
	SessionID    string          `json:"session_id"`
	TurnID       string          `json:"turn_id"`
	Transcript   string          `json:"transcript_path"`
	CWD          string          `json:"cwd"`
	EventName    string          `json:"hook_event_name"`
	ToolName     string          `json:"tool_name"`
	ToolInput    json.RawMessage `json:"tool_input"`
	ToolResponse json.RawMessage `json:"tool_response"`
	ToolCallID   string          `json:"tool_use_id"`
}

type Response struct {
	Decision           string              `json:"decision,omitempty"`
	Reason             string              `json:"reason,omitempty"`
	SystemMessage      string              `json:"systemMessage,omitempty"`
	HookSpecificOutput *HookSpecificOutput `json:"hookSpecificOutput,omitempty"`
}

type HookSpecificOutput struct {
	HookEventName     string `json:"hookEventName,omitempty"`
	AdditionalContext string `json:"additionalContext,omitempty"`
}

type OperationalError struct {
	Code    string `json:"code"`
	Message string `json:"message"`
}

type Evaluation struct {
	Findings []model.Finding
	Error    *OperationalError
}

type fileChange struct {
	Path       string
	NewPath    string
	Source     []byte
	OldText    string
	NewText    string
	PatchLines []string
	Delete     bool
}

var selectedCategories = map[model.Category]bool{
	model.CategoryOrdinary:      true,
	model.CategoryDocumentation: true,
}

func ParseMode(value string) (Mode, error) {
	mode := Mode(strings.ToLower(strings.TrimSpace(value)))
	if mode != ModeHard && mode != ModeWarn {
		return "", fmt.Errorf("unsupported hook mode %q", value)
	}
	return mode, nil
}

func Run(input io.Reader, output io.Writer, mode Mode) error {
	decoder := json.NewDecoder(input)
	var event Event
	if err := decoder.Decode(&event); err != nil {
		return writeResponse(output, responseForError(mode, "invalid_event", fmt.Sprintf("could not decode hook event: %v", err)))
	}
	response := Process(event, mode)
	return writeResponse(output, response)
}

func Process(event Event, mode Mode) Response {
	if mode != ModeHard && mode != ModeWarn {
		return responseForError(mode, "invalid_mode", fmt.Sprintf("unsupported hook mode %q", mode))
	}

	eventName := strings.ToLower(strings.TrimSpace(event.EventName))
	toolName := strings.ToLower(strings.TrimSpace(event.ToolName))
	if eventName != "pretooluse" && eventName != "pre_tool_use" {
		return Response{}
	}
	if event.CWD == "" {
		cwd, err := os.Getwd()
		if err != nil {
			return responseForError(mode, "cwd_unavailable", err.Error())
		}
		event.CWD = cwd
	}
	root, err := filepath.Abs(event.CWD)
	if err != nil {
		return responseForError(mode, "cwd_invalid", err.Error())
	}

	return processPre(event, toolName, root, mode)
}

func processPre(event Event, toolName, root string, mode Mode) Response {
	if toolName != "apply_patch" && toolName != "edit" && toolName != "write" && toolName != "write_file" && toolName != "file_write" {
		return Response{}
	}
	changes, err := reconstructChanges(event.ToolInput)
	if err != nil {
		return responseForError(mode, "proposal_unreadable", err.Error())
	}
	findings, err := evaluateChanges(root, changes)
	if err != nil {
		return responseForError(mode, "proposal_evaluation_failed", err.Error())
	}
	return responseForFindings(mode, findings)
}

func responseForFindings(mode Mode, findings []model.Finding) Response {
	if len(findings) == 0 {
		return Response{}
	}
	reason := formatFindings(findings)
	if mode == ModeHard {
		return Response{Decision: "block", Reason: reason}
	}
	guidance := reason + " Use Git history for history, pull-request descriptions for rationale, simplified code or a nearby README for complexity, and Markdown for general documentation."
	return Response{
		SystemMessage: guidance,
		HookSpecificOutput: &HookSpecificOutput{
			HookEventName:     "PreToolUse",
			AdditionalContext: guidance,
		},
	}
}

func responseForError(mode Mode, code, message string) Response {
	text := fmt.Sprintf("ban-code-comments hook operational error (%s): %s", code, message)
	if mode == ModeHard {
		return Response{Decision: "block", Reason: text}
	}
	return Response{SystemMessage: text}
}

func writeResponse(output io.Writer, response Response) error {
	encoder := json.NewEncoder(output)
	return encoder.Encode(response)
}

func formatFindings(findings []model.Finding) string {
	ordered := append([]model.Finding(nil), findings...)
	sort.SliceStable(ordered, func(i, j int) bool {
		if ordered[i].Path != ordered[j].Path {
			return ordered[i].Path < ordered[j].Path
		}
		if ordered[i].Range.Start.Line != ordered[j].Range.Start.Line {
			return ordered[i].Range.Start.Line < ordered[j].Range.Start.Line
		}
		return ordered[i].Range.Start.Column < ordered[j].Range.Start.Column
	})
	parts := make([]string, 0, len(ordered))
	for _, finding := range ordered {
		parts = append(parts, fmt.Sprintf("%s:%d:%d [%s] %s", finding.Path, finding.Range.Start.Line, finding.Range.Start.Column, finding.Category, strings.TrimSpace(finding.Text)))
	}
	return "ban-code-comments found newly introduced comments: " + strings.Join(parts, "; ")
}

func evaluateChanges(root string, changes []fileChange) ([]model.Finding, error) {
	findings := make([]model.Finding, 0)
	for _, change := range changes {
		oldPath, err := resolvePath(root, change.Path)
		if err != nil {
			return nil, err
		}
		oldSource, readErr := os.ReadFile(oldPath)
		if readErr != nil && !errors.Is(readErr, os.ErrNotExist) {
			return nil, readErr
		}
		newSource := change.Source
		if change.OldText != "" || change.NewText != "" {
			if !strings.Contains(string(oldSource), change.OldText) {
				return nil, fmt.Errorf("proposed edit could not find old text in %s", change.Path)
			}
			newSource = []byte(strings.Replace(string(oldSource), change.OldText, change.NewText, 1))
		}
		if len(change.PatchLines) > 0 {
			newSource, err = applyUpdatePatch(oldSource, change.PatchLines)
			if err != nil {
				return nil, fmt.Errorf("apply proposed patch to %s: %w", change.Path, err)
			}
		}
		if change.Delete {
			newSource = nil
		}
		oldFindingPath := displayPath(root, change.Path)
		oldFindings := findingsFor(oldFindingPath, oldSource)
		newPath := change.Path
		if change.NewPath != "" {
			if _, err := resolvePath(root, change.NewPath); err != nil {
				return nil, err
			}
			newPath = change.NewPath
		}
		newFindings := findingsFor(displayPath(root, newPath), newSource)
		findings = append(findings, newFindingsDifference(oldFindings, newFindings)...)
	}
	return findings, nil
}

func findingsFor(path string, source []byte) []model.Finding {
	language, ok := languages.Lookup(path)
	if !ok {
		return nil
	}
	return scanner.Scan(filepath.ToSlash(path), language, source, selectedCategories)
}

func newFindingsDifference(before, after []model.Finding) []model.Finding {
	counts := make(map[string]int, len(before))
	for _, finding := range before {
		counts[findingKey(finding)]++
	}
	result := make([]model.Finding, 0, len(after))
	for _, finding := range after {
		key := findingKey(finding)
		if counts[key] > 0 {
			counts[key]--
			continue
		}
		result = append(result, finding)
	}
	return result
}

func findingKey(finding model.Finding) string {
	return string(finding.Language) + "\x00" + string(finding.Category) + "\x00" + finding.Text
}

func resolvePath(root, name string) (string, error) {
	name = strings.TrimSpace(filepath.ToSlash(name))
	name = strings.TrimPrefix(name, "a/")
	name = strings.TrimPrefix(name, "b/")
	if name == "" {
		return "", fmt.Errorf("invalid proposed path %q", name)
	}
	path := filepath.Clean(filepath.Join(root, filepath.FromSlash(name)))
	if filepath.IsAbs(filepath.FromSlash(name)) {
		path = filepath.Clean(filepath.FromSlash(name))
	}
	relative, err := filepath.Rel(root, path)
	if err != nil || relative == ".." || strings.HasPrefix(relative, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("proposed path escapes workspace: %q", name)
	}
	return path, nil
}

func displayPath(root, name string) string {
	path, err := resolvePath(root, name)
	if err != nil {
		return filepath.ToSlash(name)
	}
	relative, err := filepath.Rel(root, path)
	if err != nil {
		return filepath.ToSlash(name)
	}
	return filepath.ToSlash(relative)
}

func reconstructChanges(raw json.RawMessage) ([]fileChange, error) {
	if len(raw) == 0 {
		return nil, errors.New("hook event has no tool input")
	}
	var inputText string
	if err := json.Unmarshal(raw, &inputText); err == nil {
		return parsePatch(inputText)
	}
	var input map[string]json.RawMessage
	if err := json.Unmarshal(raw, &input); err != nil {
		return nil, fmt.Errorf("tool input is not an object: %w", err)
	}
	if patch := rawString(input, "patch"); patch != "" {
		return parsePatch(patch)
	}
	for _, key := range []string{"input", "command"} {
		if value := rawString(input, key); strings.HasPrefix(strings.TrimSpace(value), "*** Begin Patch") {
			return parsePatch(value)
		}
	}
	path := firstString(input, "path", "file_path", "filePath", "filename")
	if path == "" {
		return nil, errors.New("tool input does not identify a file")
	}
	if content, ok := rawBytes(input, "content", "new_content", "newContent"); ok {
		return []fileChange{{Path: path, Source: content}}, nil
	}
	oldText, oldOK := rawStringOK(input, "old_string", "oldString")
	newText, newOK := rawStringOK(input, "new_string", "newString")
	if oldOK && newOK {
		return []fileChange{{Path: path, OldText: oldText, NewText: newText}}, nil
	}
	return nil, errors.New("tool input does not contain proposed file content")
}

func rawString(input map[string]json.RawMessage, key string) string {
	value, ok := input[key]
	if !ok {
		return ""
	}
	var result string
	if json.Unmarshal(value, &result) != nil {
		return ""
	}
	return result
}

func rawStringOK(input map[string]json.RawMessage, keys ...string) (string, bool) {
	for _, key := range keys {
		if value, ok := input[key]; ok {
			var result string
			if err := json.Unmarshal(value, &result); err == nil {
				return result, true
			}
		}
	}
	return "", false
}

func rawBytes(input map[string]json.RawMessage, keys ...string) ([]byte, bool) {
	for _, key := range keys {
		if value, ok := input[key]; ok {
			var result string
			if err := json.Unmarshal(value, &result); err == nil {
				return []byte(result), true
			}
		}
	}
	return nil, false
}

func firstString(input map[string]json.RawMessage, keys ...string) string {
	for _, key := range keys {
		if value := rawString(input, key); value != "" {
			return value
		}
	}
	return ""
}

func parsePatch(patch string) ([]fileChange, error) {
	lines := strings.Split(strings.ReplaceAll(patch, "\r\n", "\n"), "\n")
	start := 0
	if len(lines) > 0 && strings.TrimSpace(lines[0]) == "*** Begin Patch" {
		start++
	}
	changes := make([]fileChange, 0)
	for start < len(lines) {
		line := lines[start]
		if strings.TrimSpace(line) == "*** End Patch" || line == "" {
			start++
			continue
		}
		switch {
		case strings.HasPrefix(line, "*** Add File: "):
			path := strings.TrimPrefix(line, "*** Add File: ")
			body, next := collectPatchBody(lines, start+1)
			changes = append(changes, fileChange{Path: path, Source: []byte(strings.Join(stripAddedLines(body), "\n"))})
			start = next
		case strings.HasPrefix(line, "*** Delete File: "):
			path := strings.TrimPrefix(line, "*** Delete File: ")
			changes = append(changes, fileChange{Path: path, Delete: true})
			start++
		case strings.HasPrefix(line, "*** Update File: "):
			path := strings.TrimPrefix(line, "*** Update File: ")
			body, next := collectPatchBody(lines, start+1)
			newPath := ""
			if len(body) > 0 && strings.HasPrefix(body[0], "*** Move to: ") {
				newPath = strings.TrimPrefix(body[0], "*** Move to: ")
				body = body[1:]
			}
			changes = append(changes, fileChange{Path: path, NewPath: newPath, PatchLines: body, Delete: false})
			start = next
		default:
			return nil, fmt.Errorf("unsupported patch line %q", line)
		}
	}
	return changes, nil
}

func collectPatchBody(lines []string, start int) ([]string, int) {
	end := start
	for end < len(lines) {
		if strings.HasPrefix(lines[end], "*** Add File: ") || strings.HasPrefix(lines[end], "*** Delete File: ") || strings.HasPrefix(lines[end], "*** Update File: ") || strings.TrimSpace(lines[end]) == "*** End Patch" {
			break
		}
		end++
	}
	return lines[start:end], end
}

func stripAddedLines(lines []string) []string {
	result := make([]string, 0, len(lines))
	for _, line := range lines {
		if strings.HasPrefix(line, "+") {
			result = append(result, strings.TrimPrefix(line, "+"))
		}
	}
	return result
}

func applyUpdatePatch(source []byte, lines []string) ([]byte, error) {
	hasFinalNewline := bytes.HasSuffix(source, []byte("\n"))
	text := strings.TrimSuffix(string(source), "\n")
	oldLines := []string{}
	if text != "" {
		oldLines = strings.Split(text, "\n")
	}
	newLines := make([]string, 0, len(oldLines))
	cursor := 0
	seenOperation := false
	flush := func(hunk []string) error {
		if len(hunk) == 0 {
			return nil
		}
		oldBlock := make([]string, 0)
		newBlock := make([]string, 0)
		for _, line := range hunk {
			if line == "\\ No newline at end of file" || line == "" || strings.HasPrefix(line, "@@") {
				continue
			}
			if line[0] != ' ' && line[0] != '-' && line[0] != '+' {
				return fmt.Errorf("unsupported hunk line %q", line)
			}
			content := line[1:]
			switch line[0] {
			case ' ', '-':
				oldBlock = append(oldBlock, content)
			}
			switch line[0] {
			case ' ', '+':
				newBlock = append(newBlock, content)
			}
		}
		if len(oldBlock) == 0 && len(newBlock) == 0 {
			return nil
		}
		match := -1
		for index := cursor; index+len(oldBlock) <= len(oldLines); index++ {
			if equalLines(oldLines[index:index+len(oldBlock)], oldBlock) {
				match = index
				break
			}
		}
		if match < 0 {
			return fmt.Errorf("hunk context was not found")
		}
		newLines = append(newLines, oldLines[cursor:match]...)
		newLines = append(newLines, newBlock...)
		cursor = match + len(oldBlock)
		seenOperation = true
		return nil
	}
	current := make([]string, 0)
	for _, line := range lines {
		if strings.HasPrefix(line, "@@") {
			if err := flush(current); err != nil {
				return nil, err
			}
			current = current[:0]
		}
		current = append(current, line)
	}
	if err := flush(current); err != nil {
		return nil, err
	}
	if !seenOperation {
		return nil, errors.New("patch contains no file operations")
	}
	newLines = append(newLines, oldLines[cursor:]...)
	result := []byte(strings.Join(newLines, "\n"))
	if hasFinalNewline {
		result = append(result, '\n')
	}
	return result, nil
}

func equalLines(left, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	for index := range left {
		if left[index] != right[index] {
			return false
		}
	}
	return true
}

File: src/hook.js
import { promises as fs } from "node:fs";
import path from "node:path";
import { lookup } from "./languages.js";
import { CATEGORIES } from "./model.js";
import { scanSource } from "./scanner.js";

const tools = new Set(["apply_patch", "edit", "write", "write_file", "file_write"]);
const selected = new Set([CATEGORIES.ORDINARY, CATEGORIES.DOCUMENTATION]);

function errorResponse(mode, code, message) {
  const text = `ban-code-comments hook operational error (${code}): ${message}`;
  return { systemMessage: text, hookSpecificOutput: { hookEventName: "PreToolUse", additionalContext: text } };
}

function resolve(root, name) {
  const normalized = String(name || "").trim().replaceAll("\\", "/").replace(new RegExp("^(?:a|b)/"), "");
  if (!normalized) throw new Error(`invalid proposed path ${JSON.stringify(name || "")}`);
  const target = path.isAbsolute(normalized) ? path.normalize(normalized) : path.resolve(root, normalized);
  const relative = path.relative(root, target);
  if (relative === ".." || relative.startsWith(`..${path.sep}`) || path.isAbsolute(relative)) throw new Error(`proposed path escapes workspace: ${JSON.stringify(name)}`);
  return target;
}

function display(root, name) { return path.relative(root, resolve(root, name)).replaceAll(path.sep, "/"); }

function stringValue(input, ...keys) { for (const key of keys) if (typeof input?.[key] === "string") return input[key]; return ""; }

function parsePatch(patch) {
  const lines = String(patch).replaceAll("\r\n", "\n").split("\n");
  const wrapped = lines[0]?.trim() === "*** Begin Patch";
  if (wrapped && !lines.some((line) => line.trim() === "*** End Patch")) throw new Error("patch is missing end marker");
  let index = wrapped ? 1 : 0;
  const changes = [];
  while (index < lines.length) {
    const line = lines[index];
    if (!line || line.trim() === "*** End Patch") { index += 1; continue; }
    if (line.startsWith("*** Add File: ")) {
      const body = collectBody(lines, index + 1); if (body.lines.some((item) => item && !item.startsWith("+"))) throw new Error(`unsupported add-file line ${JSON.stringify(body.lines.find((item) => item && !item.startsWith("+")))}`); changes.push({ path: line.slice(14), source: body.lines.filter((item) => item.startsWith("+")).map((item) => item.slice(1)).join("\n") }); index = body.next; continue;
    }
    if (line.startsWith("*** Delete File: ")) { changes.push({ path: line.slice(17), delete: true }); index += 1; continue; }
    if (line.startsWith("*** Update File: ")) {
      const body = collectBody(lines, index + 1); let newPath = ""; let patchLines = body.lines;
      if (patchLines[0]?.startsWith("*** Move to: ")) { newPath = patchLines[0].slice(13); patchLines = patchLines.slice(1); }
      changes.push({ path: line.slice(17), newPath, patchLines }); index = body.next; continue;
    }
    throw new Error(`unsupported patch line ${JSON.stringify(line)}`);
  }
  return changes;
}

function collectBody(lines, start) {
  let next = start;
  while (next < lines.length && !/^(?:\*\*\* (?:Add|Delete|Update) File:|\*\*\* End Patch)/.test(lines[next])) next += 1;
  return { lines: lines.slice(start, next), next };
}

function reconstruct(raw) {
  if (raw === undefined || raw === null) throw new Error("hook event has no tool input");
  if (typeof raw === "string") return parsePatch(raw);
  if (typeof raw !== "object" || Array.isArray(raw)) throw new Error("tool input is not an object");
  const patchKeys = ["patch", "input", "command"].filter((key) => Object.hasOwn(raw, key));
  const contentKeys = ["content", "new_content", "newContent"].filter((key) => Object.hasOwn(raw, key));
  const pathKeys = ["path", "file_path", "filePath", "filename"].filter((key) => Object.hasOwn(raw, key));
  const editKeys = ["old_string", "oldString", "new_string", "newString"].filter((key) => Object.hasOwn(raw, key));
  if (patchKeys.length > 1 || contentKeys.length > 1 || pathKeys.length > 1) throw new Error("tool input contains duplicate field aliases");
  if (patchKeys.length > 0 && (contentKeys.length > 0 || pathKeys.length > 0 || editKeys.length > 0)) throw new Error("tool input contains conflicting proposal fields");
  if (contentKeys.length > 0 && editKeys.length > 0) throw new Error("tool input contains conflicting proposal fields");
  const patch = stringValue(raw, "patch", "input", "command");
  if (patch.trimStart().startsWith("*** Begin Patch")) return parsePatch(patch);
  const filePath = stringValue(raw, "path", "file_path", "filePath", "filename");
  if (!filePath) throw new Error("tool input does not identify a file");
  for (const key of ["content", "new_content", "newContent"]) if (typeof raw[key] === "string") return [{ path: filePath, source: raw[key] }];
  const oldKeys = ["old_string", "oldString"].filter((key) => Object.hasOwn(raw, key));
  const newKeys = ["new_string", "newString"].filter((key) => Object.hasOwn(raw, key));
  if (oldKeys.length > 1 || newKeys.length > 1) throw new Error("edit input contains duplicate field aliases");
  if (oldKeys.length > 0) {
    const oldText = raw[oldKeys[0]];
    const newText = newKeys.length > 0 ? raw[newKeys[0]] : undefined;
    if (typeof oldText !== "string") throw new Error("edit input must include string old_string");
    if (oldText.length === 0) throw new Error("edit input must include non-empty old_string");
    if (typeof newText !== "string") throw new Error("edit input must include string new_string");
    return [{ path: filePath, oldText, newText }];
  }
  throw new Error("tool input does not contain proposed file content");
}

function applyPatch(source, patchLines) {
  const finalNewline = source.endsWith("\n");
  const oldLines = source.replace(/\n$/, "").split("\n");
  if (oldLines.length === 1 && oldLines[0] === "") oldLines.length = 0;
  const output = []; let cursor = 0; let changed = false; let hunk = [];
  const flush = () => {
    const operations = hunk.filter((line) => line && !line.startsWith("@@") && line !== "\\ No newline at end of file");
    if (!operations.length) return;
    const oldBlock = operations.filter((line) => line[0] === " " || line[0] === "-").map((line) => line.slice(1));
    const newBlock = operations.filter((line) => line[0] === " " || line[0] === "+").map((line) => line.slice(1));
    if (operations.some((line) => ![" ", "+", "-"].includes(line[0]))) throw new Error(`unsupported hunk line ${JSON.stringify(operations.find((line) => ![" ", "+", "-"].includes(line[0])))}`);
    let match = -1;
    for (let candidate = cursor; candidate + oldBlock.length <= oldLines.length; candidate += 1) if (oldLines.slice(candidate, candidate + oldBlock.length).every((line, i) => line === oldBlock[i])) { match = candidate; break; }
    if (match < 0) throw new Error("hunk context was not found");
    output.push(...oldLines.slice(cursor, match), ...newBlock); cursor = match + oldBlock.length; changed = true;
  };
  for (const line of patchLines) { if (line.startsWith("@@")) { flush(); hunk = []; } hunk.push(line); }
  flush();
  if (!changed) throw new Error("patch contains no file operations");
  output.push(...oldLines.slice(cursor));
  return output.join("\n") + (finalNewline ? "\n" : "");
}

function difference(before, after) {
  const counts = new Map();
  for (const item of before) { const key = `${item.language}\0${item.category}\0${item.text}`; counts.set(key, (counts.get(key) || 0) + 1); }
  return after.filter((item) => { const key = `${item.language}\0${item.category}\0${item.text}`; const count = counts.get(key) || 0; if (count) { counts.set(key, count - 1); return false; } return true; });
}

async function findingsFor(root, filePath, source) {
  const language = lookup(filePath);
  return language ? scanSource(source, display(root, filePath), language, selected) : [];
}

async function evaluate(root, changes) {
  const findings = [];
  for (const change of changes) {
    const oldPath = resolve(root, change.path);
    let oldSource = "";
    try { oldSource = await fs.readFile(oldPath, "utf8"); } catch (error) { if (error.code !== "ENOENT") throw error; }
    let newSource = change.source;
    if (change.oldText !== undefined) { if (!oldSource.includes(change.oldText)) throw new Error(`proposed edit could not find old text in ${change.path}`); newSource = oldSource.replace(change.oldText, change.newText); }
    if (change.patchLines?.length) newSource = applyPatch(oldSource, change.patchLines);
    if (change.delete) newSource = "";
    const newPath = change.newPath ? resolve(root, change.newPath) && change.newPath : change.path;
    findings.push(...difference(await findingsFor(root, change.path, oldSource), await findingsFor(root, newPath, newSource)));
  }
  return findings;
}

function formatFindings(findings) {
  return `ban-code-comments found newly introduced comments: ${[...findings].sort((a, b) => a.path.localeCompare(b.path) || a.range.start.line - b.range.start.line || a.range.start.column - b.range.start.column).map((item) => `${item.path}:${item.range.start.line}:${item.range.start.column} [${item.category}] ${item.text.trim()}`).join("; ")}`;
}

function findingResponse(mode, findings) {
  if (!findings.length) return {};
  const reason = formatFindings(findings);
  if (mode === "hard") return { decision: "block", reason };
  const guidance = `${reason} Use Git history for history, pull-request descriptions for rationale, simplified code or a nearby README for complexity, and Markdown for general documentation.`;
  return { systemMessage: guidance, hookSpecificOutput: { hookEventName: "PreToolUse", additionalContext: guidance } };
}

export async function evaluateHook(event, mode = "hard") {
  if (mode !== "hard" && mode !== "warn") return errorResponse(mode, "invalid_mode", `unsupported hook mode ${JSON.stringify(mode)}`);
  const eventName = String(event?.hook_event_name || "").trim().toLowerCase().replaceAll("_", "");
  const toolName = String(event?.tool_name || "").trim().toLowerCase();
  if (eventName !== "pretooluse" || !tools.has(toolName)) return {};
  try { return findingResponse(mode, await evaluate(path.resolve(event.cwd || process.cwd()), reconstruct(event.tool_input))); }
  catch (error) { return errorResponse(mode, "proposal_unreadable", error.message); }
}

async function decodeInput(input) {
  if (typeof input === "string") return JSON.parse(input);
  if (Buffer.isBuffer(input)) return JSON.parse(input.toString("utf8"));
  if (input && typeof input === "object" && typeof input[Symbol.asyncIterator] === "function") {
    const chunks = [];
    for await (const chunk of input) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(String(chunk)));
    return JSON.parse(Buffer.concat(chunks).toString("utf8"));
  }
  if (input && typeof input === "object" && !Array.isArray(input)) return input;
  throw new Error("hook input is neither JSON text nor an event object");
}

export async function runHook(input, mode = "hard", output = process.stdout) {
  let event;
  try { event = await decodeInput(input); } catch (error) { output.write(`${JSON.stringify(errorResponse(mode, "invalid_event", `could not decode hook event: ${error.message}`))}\n`); return 0; }
  try { output.write(`${JSON.stringify(await evaluateHook(event, mode))}\n`); }
  catch (error) { output.write(`${JSON.stringify(errorResponse(mode, "hook_failure", error.message))}\n`); }
  return 0;
}

export { applyPatch, formatFindings, parsePatch, reconstruct, resolve };
